lru.h

```diff
@@ -120,24 +121,35 @@ template<typename F> struct ConcurrentLru {
             typename boost::remove_reference<
                 typename F::result_type>::type
             >::type const;
-        using result_type = std::shared_ptr<underlying_type>;
+        using result_type = std::shared_future<std::shared_ptr<underlying_type>>;
+
         result_type operator()(argument_type arg) const {
-            return std::make_shared<underlying_type>(f(arg));
+            //build a future that will be lazy initialized
+            return std::async(std::launch::deferred, [&](){
+                return std::make_shared<underlying_type>(f(arg));
+            }).share();
         }
     };
     Lru<SharedPtrF> lru;
     std::unique_ptr<std::mutex> mutex = std::make_unique<std::mutex>();
 
 public:
-    using result_type = typename SharedPtrF::result_type;
+    using result_type = typename std::shared_ptr<typename SharedPtrF::underlying_type>;
     using argument_type = typename SharedPtrF::argument_type;
 
     ConcurrentLru(F fun, size_t max = 10): lru(SharedPtrF{std::move(fun)}, max) {}
     ConcurrentLru(ConcurrentLru&&) = default;// needed by old version of gcc
 
     result_type operator()(argument_type arg) const {
-        std::lock_guard<std::mutex> lock(*mutex);
-        return lru(arg);
+        typename SharedPtrF::result_type future;
+        {
+            std::lock_guard<std::mutex> lock(*mutex);
+            future = lru(arg);
+        }
+        // the shared_future must never be return outside of the LRU object
+        // F can take argument arg as reference, if F is called by the future outside operator() arg can
+        // be an invalid reference
```

Comment:
As arg might be a reference, the maybe newly created future must be run before the end of the current method, else we can have a use after free.